aws/s3.go

```diff
@@ -0,0 +1,159 @@
+			exists, err := checkS3BucketExists(svc, *bucket.Name)
+			if err != nil {
+				return nil, errors.WithStackTrace(err)
+			}
+			if !exists {
+				logging.Logger.Warnf("Failed to get bucket location for - %s - ignoring as bucket might be deleted.", *bucket.Name)
+				continue
+			}
+		}
+
+		var bucketRegion string
+		if result.LocationConstraint == nil {
+			// GetBucketLocation returns nil for us-east-1
+			// https://github.com/aws/aws-sdk-go/issues/1687
+			bucketRegion = "us-east-1"
+		} else {
+			bucketRegion = *result.LocationConstraint
+		}
+
+		if _, ok := bucketNamesPerRegion[bucketRegion]; !ok {
+			bucketNamesPerRegion[bucketRegion] = []*string{}
+		}
+
+		bucketNamesPerRegion[bucketRegion] = append(bucketNamesPerRegion[bucketRegion], bucket.Name)
+	}
+
+	return bucketNamesPerRegion, nil
+}
+
+// nukeAllS3Buckets deletes all S3 buckets passed as input
+func nukeAllS3Buckets(session *session.Session, bucketNames []*string) (delCount int, err error) {
+	svc := s3.New(session)
+
+	if len(bucketNames) == 0 {
+		logging.Logger.Infof("No S3 Buckets to nuke in region %s", *session.Config.Region)
+		return 0, nil
+	}
+
+	bucketCount := 0
+	totalCount := len(bucketNames)
+
+	logging.Logger.Infof("Deleting - %d S3 Buckets in region %s", totalCount, *session.Config.Region)
+
+	for _, bucketName := range bucketNames {
+		// It might be possible that between listing the bucket and deleting the bucket - the target bucket got deleted
+		// Verify if that is the case.
+		exists, err := checkS3BucketExists(svc, *bucketName)
+		if err != nil {
+			return delCount, err
+		}
+		if !exists {
+			logging.Logger.Warnf("Failed to find bucket - %s - ignoring as bucket might be deleted.", *bucketName)
+			continue
+		}
+
+		bucketCount++
+		logging.Logger.Infof("Deleting - %d/%d - S3 Bucket: %s", bucketCount, totalCount, *bucketName)
+
+		iter := s3manager.NewDeleteListIterator(svc, &s3.ListObjectsInput{
+			Bucket: bucketName,
+		})
+
+		if err := s3manager.NewBatchDeleteWithClient(svc).Delete(aws.BackgroundContext(), iter); err != nil {
+			logging.Logger.Errorf("[Failed] - %d/%d - Bucket: %s - object deletion error - %s", bucketCount, totalCount, *bucketName, err)
+			return delCount, err
+		}
+
+		_, err = svc.DeleteBucket(&s3.DeleteBucketInput{
+			Bucket: bucketName,
+		})
+
+		if err != nil {
+			logging.Logger.Errorf("[Failed] - %d/%d - Bucket: %s - bucket deletion error - %s", bucketCount, totalCount, *bucketName, err)
+			return delCount, err
+		}
+
+		logging.Logger.Infof("Awaiting deletion - %d/%d - Bucket: %s", bucketCount, totalCount, *bucketName)
+
+		err = svc.WaitUntilBucketNotExists(&s3.HeadBucketInput{
+			Bucket: bucketName,
+		})
+
+		if err != nil {
+			logging.Logger.Errorf("[Failed] - %d/%d - Bucket: %s - waiting deletion status error - %s", bucketCount, totalCount, *bucketName, err)
+			return delCount, err
+		}
+
+		logging.Logger.Infof("[OK] - %d/%d - Bucket: %s - deleted", bucketCount, totalCount, *bucketName)
+		delCount++
```

Comment:
This whole routine (deleting a single bucket) should be extracted out into a separate function for easier readability.